### PR TITLE
[confignet] Change `Transport` from `string` to `TransportType`

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:** 
Changes `Transport` from a `string` to a new `TransportType`.  Implements `UnmarshalText` for `TransportType` to enforce values.

This PR may be too much - it introduces a breaking change a lot of new public APIs that may not be worth it for such a small module. If we don't like the surface area this creates or the breaking change, but we still want to enforce transport type values, I think implementing `Validate` keeps the API footprint smaller and isn't breaking.

**Link to tracking Issue:** <Issue number if applicable>

Closes https://github.com/open-telemetry/opentelemetry-collector/issues/9364

**Documentation:** <Describe the documentation added.>

Added godoc comments

### CODE REVIEW

File: config/confignet/confignet.go
Comment (mx-psi) [inline config/confignet/confignet.go:28]: Wondering if these should be `Unix`, `Unixgram` and `UnixPacket` instead

File: config/confignet/confignet.go
Comment (TylerHelmuth) [inline config/confignet/confignet.go:73]: This is breaking stuff in Contrib (expected).  Do we need to introduce this change as a new field called `TransportType` and follow the deprecation process to get rid of `Transport` and then rename back to `Transport`?  I'd prefer not to do another 3-release cycle change.

File: config/confignet/confignet.go
Comment (TylerHelmuth) [inline config/confignet/confignet.go:14]: If we go forward with this change, should we call this `TransportType` or `Transport`?  There was a comment [here](https://github.com/open-telemetry/opentelemetry-collector/issues/6490#issuecomment-1307663325) that maybe we shouldn't include `Type` in these names.

My opinion is to keep `Type` in the name.

Comment (TylerHelmuth) [reply config/confignet/confignet.go:14]: Related to https://github.com/open-telemetry/opentelemetry-collector/issues/9388

Comment (dmitryax) [reply config/confignet/confignet.go:14]: I think we should finalize the discussion in https://github.com/open-telemetry/opentelemetry-collector/issues/9388 before making any changes here

Comment (dmitryax) [reply config/confignet/confignet.go:14]: `Transport` sounds good to me in this particular case though

File: config/confignet/confignet.go
Comment (dmitryax) [inline config/confignet/confignet.go:28]: The constants should start with the type as we do in other components e.g. `TransportTypeUnixPacket`. It might be worth putting such guidelines in docs.

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go:14]: Wrote up a quick summary of state of names here: https://github.com/open-telemetry/opentelemetry-collector/issues/9388#issuecomment-1915344957.  I think if we standardize on not including `Type` that there are some names in pdata that become "out-of-standard".  Since pdata is 1.0 and sets the precedence, I'm definitely in favor of keeping `Type` in the name.

I also personally find it clearer at a glance.  `confignet.Transport` conveys less information (in my opinion), than `confignet.TransportType`.  I feel like `Type` in the name quickly tells me it is used for enums.

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go:28]: Totally agree.  I can document all the enum naming decisions once we decide something on the `Type` or not `Type` discussion.

File: config/confignet/confignet.go
Comment (mx-psi) [reply config/confignet/confignet.go:14]: I think we should file a separate PR for the general rule and discuss it there (I mentioned what I think we should cover on https://github.com/open-telemetry/opentelemetry-collector/issues/9388#issuecomment-1919075284, if you think there are other items to add please say so!)

Comment (TylerHelmuth) [reply config/confignet/confignet.go:14]: Based on the new expectations for enumeration names, the name needs to convey a sense of categorization and I don't feel like the word `Transport` on its own is enough. I believe `TransportType` meets our expectations better.

File: config/confignet/confignet.go
Comment (mx-psi) [inline config/confignet/confignet.go:82]: nit: I think having an explicit `String` method would be better

Comment (TylerHelmuth) [reply config/confignet/confignet.go:82]: @mx-psi `String` won't play well with the consts unless we make the function a value receiver right? But then we'd be mixing value and reference receivers.

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go]: Looking at this again and taking the new Enumeration expectations into account, I think it is better if the config field name matches the enum name.  I've pushed a commit that adds a new field, `TransportType`, and deprecates `Transport`.  If we can get this merged before 0.94.0 that means we can remove `Transport` after `0.94.0` and mark as stable for the `0.95.0` release.

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go:14]: Hmm but that breaks user configurations, not just the API.  What do you think?

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go]: Hmm but that breaks user configurations (config goes from transport to transport_type), not just the API. What do you think?

Comment (mx-psi) [reply config/confignet/confignet.go]: @TylerHelmuth we can put `transport` as the `mapstructure` tag for `TransportType` and put some dummy value on the deprecated field (we tipically have used `-` for unexported fields). Since `TransportType` unmarshals in the same way as `Transport`, this should make it non-breaking for user configurations

File: config/confignet/confignet.go
Comment (mx-psi) [reply config/confignet/confignet.go:14]: Agreed, `TransportType` seems to fit the rules better

File: config/confignet/confignet_test.go
Comment (mx-psi) [inline config/confignet/confignet_test.go:19]:
```suggestion
		TransportType: confignet.TransportTypeTCP,
```

File: config/confignet/confignet_test.go
Comment (mx-psi) [inline config/confignet/confignet_test.go:54]:
```suggestion
		TransportType: confignet.TransportTypeTCP,
```

File: receiver/otlpreceiver/testdata/uds.yaml
Comment (mx-psi) [inline receiver/otlpreceiver/testdata/uds.yaml:4]: I think we should revert this back using the technique I mentioned on a comment above

**Review by mx-psi — COMMENTED** [review]
IMO this PR should not make any changes that affect end users like changing the struct tags related to this. 

We may consider doing the breaking change on a separate PR (personally I don't think it is worth it), but since we can consider the changes separately, I think we should keep each change on a separate PR

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go]: Oh clever, I'll make that change

File: config/confignet/confignet.go
Comment (mx-psi) [reply config/confignet/confignet.go]: Yup, makes sense, we can resolve this

File: receiver/otlpreceiver/testdata/uds.yaml
Comment (TylerHelmuth) [reply receiver/otlpreceiver/testdata/uds.yaml:4]: Will do

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go]: Done

File: receiver/otlpreceiver/testdata/uds.yaml
Comment (TylerHelmuth) [reply receiver/otlpreceiver/testdata/uds.yaml]: Done

**Review by mx-psi — APPROVED** [review]

File: config/confignet/confignet.go
Comment (bogdandrutu) [inline config/confignet/confignet.go:35]: What about empty string, that will error.

File: config/confignet/confignet.go
Comment (bogdandrutu) [inline config/confignet/confignet.go:97]: Would also add a `Validate` for NetAddr and `NetTCP` since user may manually setup the transport (in code) and may be a wrong invalid value (typo).

File: config/confignet/confignet.go
Comment (bogdandrutu) [inline config/confignet/confignet.go:91]: I think there is a bug here. The reason is that implementations may set the default `Transport` to `"foo"` then user sets it to `"bar"`, based on this logic if I am not mistaken you will end up with `"foo"`.

So in order to make it work (I think we can still do it if you want to not just break the go API) with implementing the `Unmarshaler` interface and before unmarshalling set the `TransportType` to the value of `Transport` then here you can ignore old value. Something like:

```golang
func (na *NetAddr) Unmarshal(confmap.Confmap cm) error {
  na.TrasportType = TransportType(na.Trasport)
  return cm.Unmarshal(na)
}
```

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go:35]: I'll add a `transportTypeEmpty`, like `configcompression` has.

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go:97]: I'll add a `Validate` function for `NetAddr`, but `TCPAddr` doesn't allow users to set a `TransportType` so there isn't anything for us to Validate yet.

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go:93]: What I tried didn't work for cases like these: https://github.com/open-telemetry/opentelemetry-collector/blob/1a57837accdff8bf9abc769a8e49b5207b52fad2/config/configgrpc/configgrpc_test.go#L618-L624 where a user is setting the public members directly, bypassing any unmashaling.

I've reverted the code to do a check for `na.Transport` in `Dial` and `Listen`. Can you explain more how that check introduces a bug for those using `Transport` instead of `TransportType`?

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go]: @mx-psi I believe configs won't be broken, but we'll break scenarios [like this](https://github.com/open-telemetry/opentelemetry-collector/actions/runs/7878581831/job/21497142469?pr=9385) where the user is expecting `Transport` to be set instead of `TransportType`.  @bogdandrutu is this the bug you're thinking of?

Comment (TylerHelmuth) [reply config/confignet/confignet.go]: All of this is just making me want to remove `Transport` and do this as a breaking change since this is an infrequently used package and the fix is easy.

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go:97]: @bogdandrutu can you explain further?

**Comment by mx-psi** [comment]
@bogdandrutu This is waiting on your review

File: config/confignet/confignet.go
Comment (bogdandrutu) [reply config/confignet/confignet.go:97]: User in their "defaultConfig"

```
return { Transport: "unix" } // configure the deprecated value (until I change the config)
```

In the yaml file:

```
  "transport": "tcp",
``` 

Then now will Call `Dial`, which protocol will be used?

Comment (TylerHelmuth) [reply config/confignet/confignet.go]: Oh I see what you mean.  I agree that is a bug.

Your proposed fix allows the default config being overridden by user config, but since `Transport` and `TransportType` are both public, users could still explicitly set `Transport` somewhere other than in via the default config function, meaning we'd still have to check for `Transport` being set in `Dial` and `Listen`.

Since this would only be a breaking change in the API and not for end user configs, I am now very much in favor of doing this change as a breaking API change.  I've update the PR to reflect that opinion.

File: config/confignet/confignet.go
Comment (TylerHelmuth) [reply config/confignet/confignet.go:76]: Ive update the PR to do this as a breaking change.

**Comment by TylerHelmuth** [comment]
All tests are now passing except Contrib, which I'll fixed after this is merged should we choose to go forward with this breaking change approach.

**Review by mx-psi — APPROVED** [review]

**Pull request merged**
